blockchain/blockchain.go

```diff
@@ -1112,6 +1119,58 @@ func (bc *blockchain) runActions(acts block.RunnableActions, ws factory.WorkingS
 	return ws.RunActions(ctx, acts.BlockHeight(), acts.Actions())
 }
 
+func (bc *blockchain) pickAndRunActions(ctx context.Context, coinBaseSelp action.SealedEnvelope,
+	actionMap map[string][]action.SealedEnvelope, ws factory.WorkingSet) (hash.Hash32B, map[hash.Hash32B]*action.Receipt, []action.SealedEnvelope, error) {
+	if bc.sf == nil {
+		return hash.ZeroHash32B, nil, nil, errors.New("statefactory cannot be nil")
+	}
+	receipts := make(map[hash.Hash32B]*action.Receipt)
+	executedActions := make([]action.SealedEnvelope, 0)
+
+	// handle coinbase transfer
+	receipt, err := ws.RunAction(ctx, coinBaseSelp)
+	if err != nil {
+		return hash.ZeroHash32B, nil, nil, errors.Wrapf(err, "Failed to update state changes for coinbase selp %s", coinBaseSelp.Hash())
+	}
+	if receipt != nil {
+		receipts[coinBaseSelp.Hash()] = receipt
+	}
+	executedActions = append(executedActions, coinBaseSelp)
+
+	raCtx, ok := protocol.GetRunActionsCtx(ctx)
+	if !ok {
+		return hash.ZeroHash32B, nil, nil, errors.New("failed to get action context")
+	}
+	// initial action iterator
+	actionIterator := actioniterator.NewActionIterator(actionMap)
+	for {
+		nextAction, ok := actionIterator.Next()
+		if !ok {
+			break
+		}
+
+		receipt, err := ws.RunAction(ctx, nextAction)
+		if err != nil {
+			if err == action.ErrHitGasLimit {
+				continue
```

Comment:
suppose to break here. Let's write a proper unit test to cover this critical function

Reply:
After skipping act on account x at nonce i, we need to make sure we will not pick act on account x nonce i + 1.

Reply:
Ping